sky/skylet/providers/azure/node_provider.py

```diff
@@ -189,11 +191,51 @@ def create_node(self, node_config, tags, count):
             VALIDITY_TAGS = [
                 TAG_RAY_CLUSTER_NAME,
                 TAG_RAY_NODE_KIND,
-                TAG_RAY_LAUNCH_CONFIG,
                 TAG_RAY_USER_NODE_TYPE,
             ]
             filters = {tag: tags[tag] for tag in VALIDITY_TAGS if tag in tags}
-            reuse_nodes = self.stopped_nodes(filters)[:count]
+            filters_with_launch_config = copy.copy(filters)
+            if TAG_RAY_LAUNCH_CONFIG in tags:
+                filters_with_launch_config[TAG_RAY_LAUNCH_CONFIG] = tags[
+                    TAG_RAY_LAUNCH_CONFIG
+                ]
+
+            # SkyPilot: We try to use the instances with the same matching launch_config first. If
+            # there is not enough instances with matching launch_config, we then use all the
+            # instances with the same matching launch_config plus some instances with wrong
+            # launch_config.
+            reuse_nodes = _filter_nodes(filters_with_launch_config)
+            assert len(reuse_nodes) <= count, len(reuse_nodes)
+            if len(reuse_nodes) < count:
+                reuse_nodes = _filter_nodes(filters, priortized_nodes=reuse_nodes)
+
+            nodes_matching_launch_config = self.stopped_nodes(
+                filters_with_launch_config
+            )
+            nodes_matching_launch_config.sort(reverse=True)
```

Comment:
nit: Comment that Azure doesn't allow accessing launch time; and that these are a bunch of str's so we're sorting by str IDs to preserve some determinism.